source-git-sync-status/README.md

```diff
@@ -0,0 +1,124 @@
+# Verifying the sync status of source-git and dist-git repos
+
+A source-git repo is in sync with the corresponding dist-git repo, if an attempt
+to sync the former to the later (by running `packit source-git update-dist-git`)
+does not produce any change in the dist-git repo.
+
+Being out-of-sync can mean:
+
+- There are commits in source-git which were not yet transferred to dist-git.
+- There are commits in dist-git, which were not created as an update from
+  source-git, and need to be synced back to it.
+- Both of the above (source-git and dist-git history has diverged).
+
+From an operational point of view, dist-git changes will need to be synced
+back to source-git, before accepting content to be transformed from source-git
+to dist-git, in order to be able to preserve dist-git as the "blessed" place
+to store the sources. In other words: all new changes to be synced from
+source-git to dist-git should be based on a commit matching the current state
+in dist-git.
+
+In order to help with this, both update commands
+(`packit source-git update-dist-git` and `packit source-git update-source-git`)
+should refuse to update the destination repository if the history of the two
+repositories has diverged. Nevertheless, it should be possible to override
+this with a `--force` flag, to allow developers to override tooling or human
+errors and bring the repositories back in sync.
```

Comment:
Depends what do you mean by "order" :-)

`--force` just makes the command to attempt the update anyways, despite that fact that the history is diverged. This means, that the update command still can fail, due to some Git conflict for example, but that's on the user to figure out.

Should I maybe make a note, that `--force` in this case is different from `git push --force` in the sense, that it will not remove commits from the destination repo, it just tries to transform outstanding commits on top of an already diverged history.

Reply:
Or maybe it should work like `git push --force`? Would that make the life of developers easier?

Reply:
My point was to find an answer when people ask us: "both histories diverged, how do we fix that?"

I personally like that `--force` will attempt to perform the update process which should fix the case of "easy spec file update in dist-git": that one will be common I think

In case of significant divergence - there is no easy solution to that, right? So I guess "recreate" will be the way to go.

I fear that doing git-push-force could erase some work done in the target repo.

Reply:
If it's about a PR or a branch which can be rebased, than the process for getting out of a diverged state would be to rebase the source-git changes on top of the outstanding dist-git ones (after syncing back those to source-git).

But there might be cases when this might not be required. If there are no conflicting changes in the histories, I don't see there would be a point in enforcing identical histories in source-git and dist-git. So if things would happen in some other order that should be also fine. But this will not work in every case, so probably we should mostly recommend the previous process :)

Should I change or add anything to this research regarding this?

Reply:
Good point. I wouldn't include that info here to not make the document too big, especially since it's research and not the official documentation. I would prefer if we could have a followup task in future - especially after we run into the divergence ourselves with our source-git repos.

If you think it's worth having the info here, I won't stop you :) LGTM as it is now